server/store/daily_summary_store.go

```diff
@@ -64,44 +64,40 @@ func (s *pluginStore) AtomicModifyDailySummaryIndex(modify func(dsumIndex DailyS
 
 func (s *pluginStore) AtomicModifyUserDailySummarySettings(mattermostUserID string, modify func(settingToStore *DailySummarySettings) (*DailySummarySettings, error)) error {
 	return s.AtomicModifyDailySummaryIndex(func(dsumIndex DailySummaryIndex) (DailySummaryIndex, error) {
-		toChange := &DailySummarySettings{MattermostUserID: mattermostUserID}
-		for _, setting := range dsumIndex {
+		foundIndex := -1
+		for i, setting := range dsumIndex {
 			if setting.MattermostUserID == mattermostUserID {
-				toChange = setting
+				foundIndex = i
+				break
 			}
 		}
 
+		toChange := &DailySummarySettings{MattermostUserID: mattermostUserID}
+		if foundIndex != -1 {
+			toChange = dsumIndex[foundIndex]
+		}
+
 		updated, err := modify(toChange)
 		if err != nil {
 			return nil, err
 		}
 
-		result := DailySummaryIndex{}
-		found := false
-		for _, setting := range dsumIndex {
-			if setting.MattermostUserID == mattermostUserID {
-				result = append(result, updated)
-				found = true
-			} else {
-				result = append(result, setting)
-			}
+		if foundIndex != -1 {
+			result := append(dsumIndex[:foundIndex], updated)
+			return append(result, dsumIndex[foundIndex+1:]...), nil
 		}
-		if !found {
-			result = append(result, updated)
-		}
-		return result, nil
+		return append(dsumIndex, updated), nil
 	})
 
 }
 
 func (s *pluginStore) AtomicDeleteUserDailySummarySettings(mattermostUserID string) error {
 	return s.AtomicModifyDailySummaryIndex(func(dsumIndex DailySummaryIndex) (DailySummaryIndex, error) {
-		result := DailySummaryIndex{}
-		for _, setting := range dsumIndex {
-			if setting.MattermostUserID != mattermostUserID {
-				result = append(result, setting)
+		for i, setting := range dsumIndex {
+			if setting.MattermostUserID == mattermostUserID {
+				return append(dsumIndex[:i], dsumIndex[i+1:]...), nil
 			}
 		}
-		return result, nil
+		return dsumIndex, nil
```

Comment:
0/5 Should return an error when it cannot find the element to delete? Probably not, but just wondering.

Reply:
This depends on whether we store the bot in the indexes. In the `store` package we do not have access to the config, so we can't assert whether the user is the bot or not. If we don't find the user for whatever reason, I don't think we would want to cancel the caller's operation anyway.

Reply:
This is really something that we need to definitively understand and agree on though. The implementation detail of not having the bot in the indexes, versus having to check in different places if the user we are interacting with is the bot.

Reply:
Does it make sense to delete the daily summary of a bot user? If the bot user should never get a daily summary, and something tries to delete its daily summary, we would consider that an error, am I right?

Nevertheless, I also wonder on concurrency cases. If we try to delete an index that does not exist, is it an error? Or we can safely assume that the action that the user wanted to happen is done (that index is no longer there)?

Reply:
> Does it make sense to delete the daily summary of a bot user?

Depending on how we initialize the daily summary settings (right now they are not automatically initialized on connect), the bot may not be in the daily summary settings.

> Nevertheless, I also wonder on concurrency cases. If we try to delete an index that does not exist, is it an error? Or we can safely assume that the action that the user wanted to happen is done (that index is no longer there)?

As of now, this function is only called when a user disconnects, including when we disconnect the bot. If deletion in the index doesn't happen, the most we should do is log it. However, I wouldn't want it to log if it is the bot user.

We could make it so the `DisconnectUser` function that calls this function checks if it is the bot user, and only calls the relevant store methods. This would allow us to correctly return and log errors for real users.